README.txt
Installing parsedatetime
------------------------

python setup.py install

Python 3.+ is required for parsedatetime v1.+


Running Unit Tests
------------------
In the source tree do the following:

    python run_tests.py parsedatetime


Using parsedatetime
-------------------

The simple example of how to use parsedatetime is:

    import parsedatetime as pdt

    cal = pdt.Calendar()

    cal.parse("tomorrow")

More detailed examples can be found in the examples/


Documentation
-------------

The generated documentation is included by default in the docs/
directory and can also be viewed online at

    http://code-bear.com/code/parsedatetime/docs/index.html

The docs can be generated (into the docs/ directory) using either
of the two commands:

    python setup.py doc
    epydoc --html --config epydoc.conf


Notes
-----

The Calendar class has a member property named ptc which
is created during the class init method to be an instance
of parsedatetime_consts.CalendarConstants() 

This code is under some serious refactoring as now it appears
it will have more than one user.  Up until now I've been the
only person to even look at the code so bear with me please :)

History
-------

The code in parsedatetime has been implemented over the years in many
different languages (C, Clipper, Delphi) as part of different
custom/proprietary systems I've worked on.  Sadly the previous code is
not "open" in any sense of that word.

When I went to work for Open Source Applications Foundation and realized
that the Chandler project could benefit from my experience with parsing
of date/time text I decided to start from scratch and implement the
code using Python and make it truly open.

After working on the initial concept and creating something that could be
shown to the Chandler folks the code has now evolved to it's current state
with the help the Chandler folks, most especially Darshana.



CHANGES.txt
07 Sep 2009 - bear

    Created branches/python25 from current trunk to save the current code

    Converted trunk to Python 3 and also refactored how the module is structured
    so that it no longer requires import parsedatetime.parsedatetime

    Bumped version to 1.0.0 to reflect the major refactoring

07 Jan 2009 - bear

    0.8.7 release
    Apply patch submitted by Michael Lim to fix the problem parsedatetime
    was having handling dates when the month day preceeded the month
    Issue 26 http://code.google.com/p/parsedatetime/issues/detail?id=26

    Fixed TestErrors when in a local where the bad date actually returns
    a date ;)

    Checked in the TestGermanLocale unit test file missed from previous commit

20 Apr 2008 - bear

    Upating Copyright year info
    Fixing defects from Google Project page

    The comparison routine for the "failing" test was not accurate.
    The test was being flagged as failing incorrectly
    Issue 18 http://code.google.com/p/parsedatetime/issues/detail?id=18

    Added patch from Bernd Zeimetz for the German localized constants!
    http://svn.debian.org/viewsvn/*checkout*/python-modules/packages/parsedatetime/trunk/debian/patches/locale-de.dpatch
    He identifies some issues with how unicode is handled and also some other
    glitches - will have to work on them
    Issue 20 http://code.google.com/p/parsedatetime/issues/detail?id=20

    Tweaked run_tests.py to default to all tests if not given on the command line
    Removed 'this' from the list of "specials" - it was causing some grief and from the
    looks of the unit tests, not all that necessary

    Worked on bug 19 - Bernd identified that for the German locale the dayofweek check
    was being triggered for the dayoffset word "morgen" (the "mo" matched the day "morgen")
    To solve this I added a small check to make sure if the whole word being checked was
    not in the dayOffsets list, and if so not trigger.
    Issue 19 http://code.google.com/p/parsedatetime/issues/detail?id=19


28 Nov 2007 - bear

    0.8.5.1 release - removed debug code

    0.8.5 release
    bumping version to 0.8.6 in trunk

    Fixing two bugs found by Chandler QA

    Time range of "today 3:30-5pm" was actually causing a traceback.
    Added a new regex to cover this range type and a new test.

    OSAF 11299 https://bugzilla.osafoundation.org/show_bug.cgi?id=11299

    A really embarrassing for a date/time library - was actually *not*
    considering leap years when returning days in a month!
    Added tests for Feb 29th of various known leap years and also added
    a check for the daysInMonth() routine which was created to replace
    the naively simple DaysInMonthList.

    OSAF 11203 https://bugzilla.osafoundation.org/show_bug.cgi?id=11203

12 Jun 2007 - bear

    0.8.4 release
    bumping version to 0.8.5 in trunk


22 Feb 2007 - bear

    Fixed a bug reported via the code.google project page by Alexis where
    parsedatetime was not parsing day suffixes properly.  For example, the
    text "Aug 25th, 2008" would return the year as 2007 - the parser was
    not 'seeing' 2008 as a part of the expression.

    The fix was to enhance one of the "long date" regexes to handle that
    situation but yet not break the current tests - always fun for sure!

    Issue 16 http://code.google.com/p/parsedatetime/issues/detail?id=16


21 Feb 2007 - bear

    Fixed a bug Brian K. (one of the Chandler devs) found when parsing with
    the locale set to fr_FR.  The phrase "same 3 folders" was causing a key
    error inside the parser and it turns out that it's because short weekday
    names in French have a trailing '.' so "sam." was being used in the
    regular expression and the '.' was being treated as a regex symbol and
    not as a period.

    It turned out to be a simple fix - just needed to add some code to run
    re.escape over the lists before adding them to the re_values dictionary.

    Also added a TestFrenchLocale set of unit tests but made them only run
    if PyICU is found until I can build an internal locale for fr_FR.
    Issue #17 http://code.google.com/p/parsedatetime/issues/detail?id=17


14 Feb 2007 - bear

    0.8.3 release

    Minor doc string changes and other typo fixes

    Updated Copyright years

    Added a fallbackLocales=[] parameter to parsedatetime_consts init routine
    to control what locales are scanned if the default or given locale is not
    found in PyICU.
    Issue #9 http://code.google.com/p/parsedatetime/issues/detail?id=9

    While working on the regex compile-on-demand issue below, I realized that
    parsedatetime was storing the compiled regex's locally and that this would
    cause prevent parsedatetime from switching locales easily.  I've always
    wanted to make it so parsedatetime can be set to parse within a locale just
    by changing a single reference - this is one step closer to that.

    Made the regex compiles on-demand to help with performance
    Requested by the Chandler folks
    Issue #15 http://code.google.com/p/parsedatetime/issues/detail?id=15

    To test the change I ran 100 times the following code:
        for i in range(0, 100):
            c = pdc.Constants()
            p = pdt.Calendar(c)
            p = None
            c = None

    and that was measured by hotshot:

        24356 function calls (22630 primitive calls) in 0.188 CPU seconds

    after the change:

        5000 function calls in 0.140 CPU seconds

    but that doesn't test the true time as it doesn't reference any regex's
    so any time saved is deferred.  To test this I then ran before and after
    tests where I parsed the major unit test bits:

    before the change:

        80290 function calls (75929 primitive calls) in 1.055 CPU seconds

    after the change:

        55803 function calls (52445 primitive calls) in 0.997 CPU seconds

    This tells me while doing the lazy compile does save time, it's not a lot

run_tests.py
import sys, os

from unittest import TestLoader, main

class ScanningLoader(TestLoader):

    def loadTestsFromModule(self, module):
        """
        Return a suite of all tests cases contained in the given module
        """
        tests = [TestLoader.loadTestsFromModule(self,module)]

        if hasattr(module, "additional_tests"):
            tests.append(module.additional_tests())

        if hasattr(module, '__path__'):
            for dir in module.__path__:
                for file in os.listdir(dir):
                    if file.endswith('.py') and file!='__init__.py':
                        if file.lower().startswith('test'):
                            submodule = module.__name__+'.'+file[:-3]
                        else:
                            continue
                    else:
                        subpkg = os.path.join(dir,file,'__init__.py')

                        if os.path.exists(subpkg):
                            submodule = module.__name__+'.'+file
                        else:
                            continue

                    tests.append(self.loadTestsFromName(submodule))

        if len(tests)>1:
            return self.suiteClass(tests)
        else:
            return tests[0] # don't create a nested suite for only one return


if __name__ == '__main__':
    main(module='parsedatetime.tests.TestGermanLocale', testLoader=ScanningLoader())



parsedatetime/tests/TestInc.py
#!/usr/bin/env python

"""
Test Calendar.Inc() routine
"""

import unittest, time, datetime
import parsedatetime.parsedatetime as pt


  # a special compare function is used to allow us to ignore the seconds as
  # the running of the test could cross a minute boundary
def _compareResults(target, value):
    t_yr, t_mth, t_dy, t_hr, t_min, _, _, _, _ = target
    v_yr, v_mth, v_dy, v_hr, v_min, _, _, _, _ = value

    return ((t_yr == v_yr) and (t_mth == v_mth) and (t_dy == v_dy) and
            (t_hr == v_hr) and (t_min == v_min)) #and (t_wd == v_wd) and (t_yd == v_yd))

class test(unittest.TestCase):
    def setUp(self):
        self.cal = pt.Calendar()
        self.yr, self.mth, self.dy, self.hr, self.mn, self.sec, self.wd, self.yd, self.isdst = time.localtime()

    def testIncMonths(self):
        s = datetime.datetime(2006, 1, 1, 12, 0, 0)
        t = datetime.datetime(2006, 2, 1, 12, 0, 0)
        self.assertTrue(_compareResults(self.cal.inc(s, month=1).timetuple(), t.timetuple()))

        s = datetime.datetime(2006, 12, 1, 12, 0, 0)
        t = datetime.datetime(2007,  1, 1, 12, 0, 0)
        self.assertTrue(_compareResults(self.cal.inc(s, month=1).timetuple(), t.timetuple()))

        # leap year, Feb 1
        s = datetime.datetime(2008, 2, 1, 12, 0, 0)
        t = datetime.datetime(2008, 3, 1, 12, 0, 0)
        self.assertTrue(_compareResults(self.cal.inc(s, month=1).timetuple(), t.timetuple()))

        # leap year, Feb 29
        s = datetime.datetime(2008, 2, 29, 12, 0, 0)
        t = datetime.datetime(2008, 3, 29, 12, 0, 0)
        self.assertTrue(_compareResults(self.cal.inc(s, month=1).timetuple(), t.timetuple()))

        s = datetime.datetime(2006,  1, 1, 12, 0, 0)
        t = datetime.datetime(2005, 12, 1, 12, 0, 0)
        self.assertTrue(_compareResults(self.cal.inc(s, month=-1).timetuple(), t.timetuple()))

        # End of month Jan 31 to Feb - Febuary only has 28 days
        s = datetime.datetime(2006, 1, 31, 12, 0, 0)
        t = datetime.datetime(2006, 2, 28, 12, 0, 0)
        self.assertTrue(_compareResults(self.cal.inc(s, month=1).timetuple(), t.timetuple()))

        # walk thru months and make sure month increment doesn't set the day
        # to be past the last day of the new month
        # think Jan transition to Feb - 31 days to 28 days
        for m in range(1, 11):
            d = self.cal.ptc.DaysInMonthList[m - 1]
            s = datetime.datetime(2006, m, d, 12, 0, 0)

            if d > self.cal.ptc.DaysInMonthList[m]:
                d = self.cal.ptc.DaysInMonthList[m]

            t = datetime.datetime(2006, m + 1, d, 12, 0, 0)

            self.assertTrue(_compareResults(self.cal.inc(s, month=1).timetuple(), t.timetuple()))


    def testIncYears(self):
        s = datetime.datetime(2006, 1, 1, 12, 0, 0)
        t = datetime.datetime(2007, 1, 1, 12, 0, 0)
        self.assertTrue(_compareResults(self.cal.inc(s, year=1).timetuple(), t.timetuple()))

        s = datetime.datetime(2006, 1, 1, 12, 0, 0)
        t = datetime.datetime(2008, 1, 1, 12, 0, 0)
        self.assertTrue(_compareResults(self.cal.inc(s, year=2).timetuple(), t.timetuple()))

        s = datetime.datetime(2006, 12, 31, 12, 0, 0)
        t = datetime.datetime(2007, 12, 31, 12, 0, 0)
        self.assertTrue(_compareResults(self.cal.inc(s, year=1).timetuple(), t.timetuple()))

        s = datetime.datetime(2006, 12, 31, 12, 0, 0)
        t = datetime.datetime(2005, 12, 31, 12, 0, 0)
        self.assertTrue(_compareResults(self.cal.inc(s, year=-1).timetuple(), t.timetuple()))

        s = datetime.datetime(2008, 3, 1, 12, 0, 0)
        t = datetime.datetime(2009, 3, 1, 12, 0, 0)
        self.assertTrue(_compareResults(self.cal.inc(s, year=1).timetuple(), t.timetuple()))

        s = datetime.datetime(2008, 3, 1, 12, 0, 0)
        t = datetime.datetime(2007, 3, 1, 12, 0, 0)
        self.assertTrue(_compareResults(self.cal.inc(s, year=-1).timetuple(), t.timetuple()))

if __name__ == "__main__":
    unittest.main()


parsedatetime/tests/__init__.py
"""
Unit tests for parsedatetime

The tests can be run as a C{suite} by running::

    python run_tests.py parsedatetime
"""

license = """
Copyright (c) 2004-2006 Mike Taylor
Copyright (c) 2006 Darshana Chhajed
All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""



parsedatetime/pdt_locales.py

"""
pdt_locales

All of the included locale classes shipped with pdt.
"""

__author__       = 'Mike Taylor (bear@code-bear.com)'
__copyright__    = 'Copyright (c) 2004 Mike Taylor'
__license__      = 'Apache v2.0'
__version__      = '1.0.0'
__contributors__ = [ 'Darshana Chhajed',
                     'Michael Lim (lim.ck.michael@gmail.com)',
                     'Bernd Zeimetz (bzed@debian.org)',
                   ]


import datetime

try:
   import PyICU as pyicu
except:
   pyicu = None


def lcase(x):
    return x.lower()


class pdtLocale_base(object):
    """
    default values for Locales
    """
    locale_keys = [ 'MonthOffsets', 'Months', 'WeekdayOffsets', 'Weekdays', 
                    'dateFormats', 'dateSep', 'dayOffsets', 'dp_order', 
                    'localeID', 'meridian', 'Modifiers', 're_sources', 're_values', 
                    'shortMonths', 'shortWeekdays', 'timeFormats', 'timeSep', 'units', 
                    'uses24', 'usesMeridian' ]

    def __init__(self):
        self.localeID      = None   # don't use a unicode string
        self.dateSep       = [ '/', '.' ]
        self.timeSep       = [ ':' ]
        self.meridian      = [ 'AM', 'PM' ]
        self.usesMeridian  = True
        self.uses24        = True

        self.WeekdayOffsets = {}
        self.MonthOffsets   = {}

        # always lowercase any lookup values - helper code expects that
        self.Weekdays      = [ 'monday', 'tuesday', 'wednesday',
                               'thursday', 'friday', 'saturday', 'sunday',
                             ]
        self.shortWeekdays = [ 'mon', 'tues', 'wed',
                               'th', 'fri', 'sat', 'sun',
                             ]
        self.Months        = [ 'january', 'february', 'march',
                               'april',   'may',      'june',
                               'july',    'august',   'september',
                               'october', 'november', 'december',
                             ]
        self.shortMonths   = [ 'jan', 'feb', 'mar',
                               'apr', 'may', 'jun',
                               'jul', 'aug', 'sep',
                               'oct', 'nov', 'dec',
                             ]
        # use the same formats as ICU by default
        self.dateFormats   = { 'full':   'EEEE, MMMM d, yyyy',
                               'long':   'MMMM d, yyyy',
                               'medium': 'MMM d, yyyy',
                               'short':  'M/d/yy',
                             }
        self.timeFormats   = { 'full':   'h:mm:ss a z',
                               'long':   'h:mm:ss a z',
                               'medium': 'h:mm:ss a',
                               'short':  'h:mm a',
                             }

        self.dp_order = [ 'm', 'd', 'y' ]

          # this will be added to re_values later
        self.units = { 'seconds': [ 'second', 'sec' ],
                       'minutes': [ 'minute', 'min' ],
                       'hours':   [ 'hour',   'hr'  ],
                       'days':    [ 'day',    'dy'  ],
                       'weeks':   [ 'week',   'wk'  ],
                       'months':  [ 'month',  'mth' ],
                       'years':   [ 'year',   'yr'  ],
                     }

          # text constants to be used by later regular expressions
        self.re_values     = { 'specials':       'in|on|of|at',
                               'timeseperator':  ':',
                               'rangeseperator': '-',
                               'daysuffix':      'rd|st|nd|th',
                               'meridian':       'am|pm|a.m.|p.m.|a|p',
                               'qunits':         'h|m|s|d|w|m|y',
                               'now':            [ 'now' ],
                             }

          # Used to adjust the returned date before/after the source
        self.Modifiers = { 'from':      1,
                           'before':   -1,
                           'after':     1,
                           'ago':      -1,
                           'prior':    -1,
                           'prev':     -1,
                           'last':     -1,
                           'next':      1,
                           'previous': -1,
                           'in a':      2,
                           'end of':    0,
                           'eod':       0,
                           'eo':        0
                         }

        self.dayOffsets = { 'tomorrow':   1,
                            'today':      0,
                            'yesterday': -1,
                          }

          # special day and/or times, i.e. lunch, noon, evening
          # each element in the dictionary is a dictionary that is used
          # to fill in any value to be replace - the current date/time will
          # already have been populated by the method buildSources
        self.re_sources    = { 'noon':      { 'hr': 12, 'mn': 0, 'sec': 0 },
                               'lunch':     { 'hr': 12, 'mn': 0, 'sec': 0 },
                               'morning':   { 'hr':  6, 'mn': 0, 'sec': 0 },
                               'breakfast': { 'hr':  8, 'mn': 0, 'sec': 0 },
                               'dinner':    { 'hr': 19, 'mn': 0, 'sec': 0 },
                               'evening':   { 'hr': 18, 'mn': 0, 'sec': 0 },
                               'midnight':  { 'hr':  0, 'mn': 0, 'sec': 0 },
                               'night':     { 'hr': 21, 'mn': 0, 'sec': 0 },
                               'tonight':   { 'hr': 21, 'mn': 0, 'sec': 0 },
                               'eod':       { 'hr': 17, 'mn': 0, 'sec': 0 },
                             }


class pdtLocale_icu(pdtLocale_base):
    """
    Create a locale from pyICU
    """
    def __init__(self, localeID):
        super( pdtLocale_icu, self ).__init__()

        self.icu = None

        if pyicu is not None:
            self.icu = pyicu.Locale(localeID)

        if self.icu is not None:
            self.symbols = pyicu.DateFormatSymbols(self.icu)

              # grab ICU list of weekdays, skipping first entry which
              # is always blank
            wd  = map(lcase, self.symbols.getWeekdays()[1:])
            swd = map(lcase, self.symbols.getShortWeekdays()[1:])

              # store them in our list with Monday first (ICU puts Sunday first)
            self.Weekdays      = wd[1:] + wd[0:1]
            self.shortWeekdays = swd[1:] + swd[0:1]
            self.Months        = map(lcase, self.symbols.getMonths())
            self.shortMonths   = map(lcase, self.symbols.getShortMonths())

            self.icu_df      = { 'full':   pyicu.DateFormat.createDateInstance(pyicu.DateFormat.kFull,   self.icu),
                                 'long':   pyicu.DateFormat.createDateInstance(pyicu.DateFormat.kLong,   self.icu),
                                 'medium': pyicu.DateFormat.createDateInstance(pyicu.DateFormat.kMedium, self.icu),
                                 'short':  pyicu.DateFormat.createDateInstance(pyicu.DateFormat.kShort,  self.icu),
                               }
            self.icu_tf      = { 'full':   pyicu.DateFormat.createTimeInstance(pyicu.DateFormat.kFull,   self.icu),
                                 'long':   pyicu.DateFormat.createTimeInstance(pyicu.DateFormat.kLong,   self.icu),
                                 'medium': pyicu.DateFormat.createTimeInstance(pyicu.DateFormat.kMedium, self.icu),
                                 'short':  pyicu.DateFormat.createTimeInstance(pyicu.DateFormat.kShort,  self.icu),
                               }
            self.dateFormats = { 'full':   self.icu_df['full'].toPattern(),
                                 'long':   self.icu_df['long'].toPattern(),
                                 'medium': self.icu_df['medium'].toPattern(),
                                 'short':  self.icu_df['short'].toPattern(),
                               }
            self.timeFormats = { 'full':   self.icu_tf['full'].toPattern(),
                                 'long':   self.icu_tf['long'].toPattern(),
                                 'medium': self.icu_tf['medium'].toPattern(),
                                 'short':  self.icu_tf['short'].toPattern(),
                               }

            am = ''
            pm = ''
            ts = ''

              # ICU doesn't seem to provide directly the date or time seperator
              # so we have to figure it out
            o = self.icu_tf['short']
            s = self.timeFormats['short']

            self.usesMeridian = 'a' in s
            slef.uses24       = 'H' in s

              # '11:45 AM' or '11:45'
            s = o.format(datetime.datetime(2003, 10, 30, 11, 45))

              # ': AM' or ':'
            s = s.replace('11', '').replace('45', '')

            if len(s) > 0:
               ts = s[0]

            if self.usesMeridian:
                 # '23:45 AM' or '23:45'
               am = s[1:].strip()
               s  = o.format(datetime.datetime(2003, 10, 30, 23, 45))

               if self.uses24:
                   s = s.replace('23', '')
               else:
                   s = s.replace('11', '')

                 # 'PM' or ''
               pm = s.replace('45', '').replace(ts, '').strip()

            self.timeSep  = [ ts ]
            self.meridian = [ am, pm ]

            o = self.icu_df['short']
            s = o.format(datetime.datetime(2003, 10, 30, 11, 45))
            s = s.replace('10', '').replace('30', '').replace('03', '').replace('2003', '')

            if len(s) > 0:
               ds = s[0]
            else:
               ds = '/'

            self.dateSep = [ ds ]
            s            = self.dateFormats['short']
            l            = s.lower().split(ds)
            dp_order     = []

            for s in l:
               if len(s) > 0:
                   dp_order.append(s[:1])

            self.dp_order = dp_order



class pdtLocale_en(pdtLocale_base):
    """
    en_US Locale
    """
    def __init__(self):
        super( pdtLocale_en, self ).__init__()

        self.localeID = 'en_US'  # don't use a unicode string
        self.uses24   = False


class pdtLocale_au(pdtLocale_base):
    """
    en_AU Locale
    """
    def __init__(self):
        super( pdtLocale_au, self ).__init__()

        self.localeID = 'en_A'   # don't use a unicode string
        self.dateSep  = [ '-', '/' ]
        self.uses24   = False

        self.dateFormats['full']   = 'EEEE, d MMMM yyyy'
        self.dateFormats['long']   = 'd MMMM yyyy'
        self.dateFormats['medium'] = 'dd/MM/yyyy'
        self.dateFormats['short']  = 'd/MM/yy'

        self.timeFormats['long']   = self.timeFormats['full']

        self.dp_order = [ 'd', 'm', 'y' ]

class pdtLocale_es(pdtLocale_base):
    """
    es Locale

    Note that I don't speak Spanish so many of the items below are still in English
    """
    def __init__(self):
        super( pdtLocale_es, self ).__init__()

        self.localeID     = 'es'   # don't use a unicode string
        self.dateSep      = [ '/' ]
        self.usesMeridian = False
        self.uses24       = True

        self.Weekdays      = [ 'lunes', 'martes', 'mi\xe9rcoles',
                               'jueves', 'viernes', 's\xe1bado', 'domingo',
                             ]
        self.shortWeekdays = [ 'lun', 'mar', 'mi\xe9',
                               'jue', 'vie', 's\xe1b', 'dom',
                             ]
        self.Months        = [ 'enero', 'febrero', 'marzo',
                               'abril', 'mayo', 'junio',
                               'julio', 'agosto', 'septiembre',
                               'octubre', 'noviembre', 'diciembre'
                             ]
        self.shortMonths   = [ 'ene', 'feb', 'mar',
                               'abr', 'may', 'jun',
                               'jul', 'ago', 'sep',
                               'oct', 'nov', 'dic'
                             ]
        self.dateFormats['full']   = "EEEE d' de 'MMMM' de 'yyyy"
        self.dateFormats['long']   = "d' de 'MMMM' de 'yyyy"
        self.dateFormats['medium'] = "dd-MMM-yy"
        self.dateFormats['short']  = "d/MM/yy"

        self.timeFormats['full']   = "HH'H'mm' 'ss z"
        self.timeFormats['long']   = "HH:mm:ss z"
        self.timeFormats['medium'] = "HH:mm:ss"
        self.timeFormats['short']  = "HH:mm"

        self.dp_order = [ 'd', 'm', 'y' ]


class pdtLocale_de(pdtLocale_base):
    """
    de_DE Locale constants

    Contributed by Debian parsedatetime package maintainer Bernd Zeimetz <bzed@debian.org>
    """
    def __init__(self):
        super( pdtLocale_de, self ).__init__()

        self.localeID      = 'de_DE'   # don't use a unicode string
        self.dateSep       = [ '.' ]
        self.timeSep       = [ ':' ]
        self.meridian      = [ ]
        self.usesMeridian  = False
        self.uses24        = True

        self.Weekdays      = [ 'montag', 'dienstag', 'mittwoch',
                               'donnerstag', 'freitag', 'samstag', 'sonntag',
                             ]
        self.shortWeekdays = [ 'mo', 'di', 'mi',
                               'do', 'fr', 'sa', 'so',
                             ]
        self.Months        = [ 'januar',  'februar',  'm\xe4rz',
                               'april',   'mai',      'juni',
                               'juli',    'august',   'september',
                               'oktober', 'november', 'dezember',
                             ]
        self.shortMonths   = [ 'jan', 'feb', 'mrz',
                               'apr', 'mai', 'jun',
                               'jul', 'aug', 'sep',
                               'okt', 'nov', 'dez',
                             ]
        self.dateFormats['full']   = 'EEEE, d. MMMM yyyy'
        self.dateFormats['long']   = 'd. MMMM yyyy'
        self.dateFormats['medium'] = 'dd.MM.yyyy'
        self.dateFormats['short']  = 'dd.MM.yy'

        self.timeFormats['full']   = 'HH:mm:ss v'
        self.timeFormats['long']   = 'HH:mm:ss z'
        self.timeFormats['medium'] = 'HH:mm:ss'
        self.timeFormats['short']  = 'HH:mm'

        self.dp_order = [ 'd', 'm', 'y' ]

        self.units['seconds'] = [ 'sekunden', 'sek',  's' ]
        self.units['minutes'] = [ 'minuten',  'min' , 'm' ]
        self.units['hours']   = [ 'stunden',  'std',  'h' ]
        self.units['days']    = [ 'tage',     't' ]
        self.units['weeks']   = [ 'wochen',   'w' ]
        self.units['months']  = [ 'monate' ]  #the short version would be a capital M,
                                              #as I understand it we can't distinguis
                                              #between m for minutes and M for months.
        self.units['years']   = [ 'jahre',    'j' ]

        self.re_values['specials']       = 'am|dem|der|im|in|den|zum'
        self.re_values['timeseperator']  = ':'
        self.re_values['rangeseperator'] = '-'
        self.re_values['daysuffix']      = ''
        self.re_values['qunits']         = 'h|m|s|t|w|m|j'
        self.re_values['now']            = [ 'jetzt' ]

          # Used to adjust the returned date before/after the source
          #still looking for insight on how to translate all of them to german.
        self.Modifiers['from']        =  1
        self.Modifiers['before']      = -1
        self.Modifiers['after']       =  1
        self.Modifiers['vergangener'] = -1
        self.Modifiers['vorheriger']  = -1
        self.Modifiers['prev']        = -1
        self.Modifiers['letzter']     = -1
        self.Modifiers['n\xe4chster'] =  1
        self.Modifiers['dieser']      =  0
        self.Modifiers['previous']    = -1
        self.Modifiers['in a']        =  2
        self.Modifiers['end of']      =  0
        self.Modifiers['eod']         =  0
        self.Modifiers['eo']          =  0

          #morgen/abermorgen does not work, see http://code.google.com/p/parsedatetime/issues/detail?id=19
        self.dayOffsets['morgen']        =  1
        self.dayOffsets['heute']         =  0
        self.dayOffsets['gestern']       = -1
        self.dayOffsets['vorgestern']    = -2
        self.dayOffsets['\xfcbermorgen'] =  2

          # special day and/or times, i.e. lunch, noon, evening
          # each element in the dictionary is a dictionary that is used
          # to fill in any value to be replace - the current date/time will
          # already have been populated by the method buildSources
        self.re_sources['mittag']         = { 'hr': 12, 'mn': 0, 'sec': 0 }
        self.re_sources['mittags']        = { 'hr': 12, 'mn': 0, 'sec': 0 }
        self.re_sources['mittagessen']    = { 'hr': 12, 'mn': 0, 'sec': 0 }
        self.re_sources['morgen']         = { 'hr':  6, 'mn': 0, 'sec': 0 }
        self.re_sources['morgens']        = { 'hr':  6, 'mn': 0, 'sec': 0 }
        self.re_sources['fr\e4hst\xe4ck'] = { 'hr':  8, 'mn': 0, 'sec': 0 }
        self.re_sources['abendessen']     = { 'hr': 19, 'mn': 0, 'sec': 0 }
        self.re_sources['abend']          = { 'hr': 18, 'mn': 0, 'sec': 0 }
        self.re_sources['abends']         = { 'hr': 18, 'mn': 0, 'sec': 0 }
        self.re_sources['mitternacht']    = { 'hr':  0, 'mn': 0, 'sec': 0 }
        self.re_sources['nacht']          = { 'hr': 21, 'mn': 0, 'sec': 0 }
        self.re_sources['nachts']         = { 'hr': 21, 'mn': 0, 'sec': 0 }
        self.re_sources['heute abend']    = { 'hr': 21, 'mn': 0, 'sec': 0 }
        self.re_sources['heute nacht']    = { 'hr': 21, 'mn': 0, 'sec': 0 }
        self.re_sources['feierabend']     = { 'hr': 17, 'mn': 0, 'sec': 0 }

